staging/src/k8s.io/apiserver/pkg/server/connectivity_service.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"bufio"
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"k8s.io/klog"
+	"net"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/apiserver/pkg/apis/apiserver"
+)
+
+var directDialer utilnet.DialFunc = (&net.Dialer{
+	Timeout:   30 * time.Second,
+	KeepAlive: 30 * time.Second,
+	DualStack: true,
+}).DialContext
+
+type ConnectivityService struct {
+	m  map[string]utilnet.DialFunc
+	mu sync.Mutex
+}
+
+type NetworkContext struct {
+	// ConnectivityServiceName is the unique name of the
+	// ConnectivityServiceConfiguration which determines
+	// the network we route the traffic to.
+	ConnectivityServiceName string
+}
+
+type ConnectivityServiceLookup func(networkContext NetworkContext) (utilnet.DialFunc, error)
+
+func createConnectDialer(connection apiserver.Connection) (utilnet.DialFunc, error) {
+	clientCert := connection.ClientCertFile
+	clientKey := connection.ClientKeyFile
+	caCert := connection.CABundle
+	proxyURL, err := url.Parse(connection.URL)
+	if err != nil {
+		return nil, fmt.Errorf("invalid proxy server url %q: %v", connection.URL, err)
+	}
+	proxyAddress := proxyURL.Host
+
+	clientCerts, err := tls.LoadX509KeyPair(clientCert, clientKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read key pair %s & %s, got %v", clientCert, clientKey, err)
+	}
+	certPool := x509.NewCertPool()
+	certBytes, err := ioutil.ReadFile(caCert)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read cert file %s, got %v", caCert, err)
+	}
+	ok := certPool.AppendCertsFromPEM(certBytes)
+	if !ok {
+		return nil, fmt.Errorf("failed to append CA cert to the cert pool")
+	}
+	contextDialer := func(ctx context.Context, network, addr string) (net.Conn, error) {
+		klog.Infof("Sending request to %q.", addr)
+		proxyConn, err := tls.Dial("tcp", proxyAddress,
+			&tls.Config{
+				ServerName:   "kubernetes-master",
```

Comment:
As I understand it, when I make a request to the server, this indicates to the server to respond with a cert signed for this name.  Is there a canonical, "this is not real" value?

Reply:
Its actually the SNI field as per https://tools.ietf.org/html/rfc4366#section-3.1. I should not have included it until we have a concrete use case. It may become useful if we ever want to do things like run multiple virtual proxy servers inside the same instance. For now I'm just going to remove it.